### PR TITLE
Perf graphing

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY

This PR aims to provide a callback plugin that can be used for local controller based performance tracking and graphing.

The PR will provide high water marks in the recap output, as well as CSV files that can be utilized to show/graph performance progression.

Currently supports:
* CPU
* Memory
* PID Count

Additionally there is a script that utilizes matplotlib for graphing the results.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/plugins/callback/cgroup_perf_recap.py

##### ANSIBLE VERSION

```paste below
2.8
```

##### ADDITIONAL INFORMATION
N/A

### CODE REVIEW

**Comment by sivel** [comment]
Wrapper script to use this callback in Tower (until Tower supports this natively):

```bash
#!/bin/bash
#
# ansible-playbook wrapper to use the cgroup_perf_recap callback plugin
# within Ansible Tower
#
# Create the cgroup control group with:
#    sudo cgcreate -a awx:awx -t awx:awx -g pids,cpuacct,memory:ansible_profile
#
# Place this file at:
#     /usr/local/bin/ansible-playbook
# and set as executable:
#     sudo chmod +x /usr/local/bin/ansible-playbook
#
# To have the cgroup_perf_recap callback write outside of the proot, you will need
# to disable "Job Isolation" or allow the bubblewrap access to other directories.
# See https://docs.ansible.com/ansible-tower/latest/html/userguide/security.html

set -e

export ANSIBLE_CALLBACK_WHITELIST=cgroup_perf_recap,${ANSIBLE_CALLBACK_WHITELIST}
export CGROUP_CONTROL_GROUP=ansible_profile
export CGROUP_CSV_OUTPUT_DIR=/tmp/%s

cgexec -g pids,cpuacct,memory:ansible_profile /usr/bin/ansible-playbook "$@"
```

**Comment by sivel** [comment]
Example graph produced from this callback/data https://user-images.githubusercontent.com/39340/47583126-16d26900-d91c-11e8-9ef5-e0b793c99e66.png

File: lib/ansible/plugins/callback/cgroup_perf_recap.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+        ini:
+          - section: callback_cgroup_perf_recap
+            key: output_format
+        type: str
+        default: csv
+        choices:
+          - csv
+          - json
+      output_dir:
+        description: Output director for files containing recorded performance readings. If the value contains a
```
Comment (chrismeyersfsu) [inline lib/ansible/plugins/callback/cgroup_perf_recap.py]: Spelling error: directory -> directory. Also should mention that the directory will be created if it doesn't exist (but the path leading up to it won't be created).

File: lib/ansible/plugins/callback/cgroup_perf_recap.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+            key: display_recap
+        type: bool
+        default: true
+      write_files:
+        description: Dictates whether files will be written containing performance readings
+        env:
+          - name: CGROUP_WRITE_FILES
+        ini:
+          - section: callback_cgroup_perf_recap
+            key: write_csv
```
Comment (chrismeyersfsu) [inline lib/ansible/plugins/callback/cgroup_perf_recap.py]: `key: write_csv` should be `key: write_files`

File: lib/ansible/plugins/callback/cgroup_perf_recap.py (hunk 1 above)
Comment (sivel) [reply lib/ansible/plugins/callback/cgroup_perf_recap.py]: Resolved

File: lib/ansible/plugins/callback/cgroup_perf_recap.py (hunk 2 above)
Comment (sivel) [reply lib/ansible/plugins/callback/cgroup_perf_recap.py]: Fixed

**Comment by chrismeyersfsu** [comment]
For runner integration, it would be helpful to output per-task-event performance data files. i.e.

```
│       │   ├── 1-34437b34-addd-45ae-819a-4d8c9711e191-cpu.json
│       │   ├── 1-34437b34-addd-45ae-819a-4d8c9711e191-mem.json
│       │   ├── 2-8c164553-8573-b1e0-76e1-00000000000d-cpu.json
│       │   ├── 2-8c164553-8573-b1e0-76e1-00000000000d-mem.json
│       │   ├── 3-8c164553-8573-b1e0-76e1-000000000008-cpu.json
│       │   ├── 3-8c164553-8573-b1e0-76e1-000000000008-mem.json
```
I would like to be able to define the filename pattern as `{counter}-{taskuuid}-{prof_type}`. This would be easy enough with python `format()`. The reason for the counter is to be able to process the files in order as they were ran. On the other hand, if we are post-processing and stitching events back together, we may want to more easily find a specific task, in which case we would want to format the files as such `{taskuuid}-{prof_type}`

**Comment by sivel** [comment]
> For runner integration, it would be helpful to output per-task-event performance data files

I'll look into this.  As of now we only open the files at the start. So I may need to add at least 1 new function, and allow this behavior to be toggled.

**Comment by sivel** [comment]
| Parameter                    | Choices/Defaults     | Configuration                                                                                                     | Comments                                                                                                                                                                                                                                                                                  |
|------------------------------|----------------------|-------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| control\_group<br>(required) |                      | [callback\_cgroup\_perf\_recap]<br>control\_group = VALUE<br><br>env:CGROUP\_CONTROL\_GROUP                       | Name of cgroups control group                                                                                                                                                                                                                                                             |
| cpu\_poll\_interval          | 0.25                 | [callback\_cgroup\_perf\_recap]<br>cpu\_poll\_interval = 0.25<br><br>env:CGROUP\_CPU\_POLL\_INTERVAL              | Interval between CPU polling for determining CPU usage. A lower value may produce inaccurate results, a higher value may not be short enough to collect results for short tasks.                                                                                                          |
| display\_recap               | (yes)/no             | [callback\_cgroup\_perf\_recap]<br>display\_recap = yes<br><br>env:CGROUP\_DISPLAY\_RECAP                         | Controls whether the recap is printed at the end, useful if you will automatically process the output files                                                                                                                                                                               |
| file\_name\_format           | %(feature)s.%(ext)s  | [callback\_cgroup\_perf\_recap]<br>file\_name\_format = %(feature)s.%(ext)s<br><br>env:CGROUP\_FILE\_NAME\_FORMAT | Format of filename. Accepts C(%(counter)s), C(%(task\_uuid)s), C(%(feature)s), C(%(ext)s). Defaults to C(%(feature)s.%(ext)s) when C(file\_per\_task) is C(False) and C(%(counter)s-%(task\_uuid)s-%(feature)s.%(ext)s) when C(True)                                                      |
| file\_per\_task              | yes/(no)             | [callback\_cgroup\_perf\_recap]<br>file\_per\_task = no<br><br>env:CGROUP\_FILE\_PER\_TASK                        | When set as C(True) along with C(write\_files), this callback will write 1 file per task instead of 1 file for the entire playbook run                                                                                                                                                    |
| output\_dir                  | /tmp/ansible-perf-%s | [callback\_cgroup\_perf\_recap]<br>output\_dir = /tmp/ansible-perf-%s<br><br>env:CGROUP\_OUTPUT\_DIR              | Output directory for files containing recorded performance readings. If the value contains a single %s, the start time of the playbook run will be inserted in that space. Only the deepest level directory will be created if it does not exist, parent directories will not be created. |
| output\_format               | (csv)/json           | [callback\_cgroup\_perf\_recap]<br>output\_format = csv<br><br>env:CGROUP\_OUTPUT\_FORMAT                         | Output format, either CSV or JSON-seq                                                                                                                                                                                                                                                     |
| write\_files                 | yes/(no)             | [callback\_cgroup\_perf\_recap]<br>write\_files = no<br><br>env:CGROUP\_WRITE\_FILES                              | Dictates whether files will be written containing performance readings                                                                                                                                                                                                                    |

**Comment by chrismeyersfsu** [comment]
kicked the tires on the per-file event, works :+1:

**Comment by bcoca** [comment]
rebuild_merge

**Pull request merged**